app/src/main/java/org/sil/gatherwords/SessionActivity.java

```diff
@@ -118,18 +129,127 @@ public void save_settings_fab_pressed(View view) {
         startActivity(i);
     }
 
-    private static class InsertSessionsTask extends AsyncTask<Session, Void, Void> {
+    private static class InsertSessionsTask extends AsyncTask<Session, Double, Void> {
         private SessionDao sDAO;
+        private WordDao wordDao;
+        // TODO: fix null object reference so we don't risk a data leak with the context
+//        private final ThreadLocal<SessionActivity> sessionActivity = new ThreadLocal<>();
+        private Long sessionID;
+        private int maxProgress;
+        private AppDatabase db;
+        private String wordList;
+        private final Context context;
 
-        InsertSessionsTask(AppDatabase db) {
+        InsertSessionsTask(SessionActivity activity) {
+//            sessionActivity.set(activity);
+//            db = AppDatabase.get(sessionActivity.get().getApplicationContext());
+            context = activity.getApplicationContext();
+            db = AppDatabase.get(context);
+            wordList = activity.worldListToLoad;
+            wordDao = db.wordDao();
             sDAO = db.sessionDao();
         }
 
         @Override
         protected Void doInBackground(Session... sessions) {
-            sDAO.insertSession(sessions);
+            List<Long> ids = sDAO.insertSession(sessions);
+
+            if (ids.get(0) != null) {
+                sessionID = ids.get(0);
+                insertFromAsset();
+            }
             return null;
         }
+
+        /**
+         * Update the main thread (the UI) with some indication of progress
+         */
+        @Override
+        protected void onProgressUpdate(Double... values) {
```

Comment:
Since it sounds you estimate this to be less than a second, I do not believe creating a progress bar will be necessary.  The word lists are unlikely to be much larger than what we have been supplied with since they are expected to be able to be completed during a single session.

I would recommend just dropping this function.